### PR TITLE
Add rsn backbone, head and pre-processing

### PR DESCRIPTION
**Pull request by wusize**

### CODE REVIEW

**Comment by jin-s13** [comment]
need unit-test for top_down_transform.py

File: mmpose/models/keypoint_heads/top_down_mspn_head.py [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+            feature, size=self.out_shape, mode='bilinear', align_corners=True)
+        if self.use_prm:
+            output = self.prm(output)
+        return output
+
+
+class PRM(nn.Module):
+    """Pose Refine Machine.
+
+    For more details about PRM, refer to Learning Delicate
```
Comment (jin-s13) [inline mmpose/models/keypoint_heads/top_down_mspn_head.py:65]: add "" to the name of the paper

Comment (wusize) [reply mmpose/models/keypoint_heads/top_down_mspn_head.py:65]: OK

File: mmpose/models/backbones/rsn.py [hunk 2]
```diff
@@ -0,0 +1,562 @@
...
+                 out_channels,
+                 num_steps=4,
+                 stride=1,
+                 downsample=None,
+                 with_cp=False,
+                 norm_cfg=dict(type='BN')):
+        super().__init__()
+        assert num_steps > 1
+        self.in_channels = in_channels
+        self.branch_channels = self.in_channels * 26 // 64
```
Comment (jin-s13) [inline mmpose/models/backbones/rsn.py:44]: Why 26? Maybe use a variable.

File: mmpose/models/backbones/rsn.py [hunk 3]
```diff
@@ -0,0 +1,562 @@
...
+    def __init__(self,
+                 block,
+                 num_blocks,
+                 num_steps=4,
+                 num_units=4,
+                 has_skip=False,
+                 norm_cfg=dict(type='BN')):
+        super().__init__()
+        self.has_skip = has_skip
+        self.in_channels = 64
```
Comment (jin-s13) [inline mmpose/models/backbones/rsn.py:150]: Can we make `in_channels` an input variable?

Comment (wusize) [reply mmpose/models/backbones/rsn.py:150]: Yes. in_channels represents the number of channels of the feature from ResNet_top.

File: mmpose/datasets/pipelines/top_down_transform.py [hunk 4]
```diff
@@ -307,14 +316,112 @@ def _generate_target(self, cfg, joints_3d, joints_3d_visible):
...
+
+        num_joints = cfg['num_joints']
+        image_size = cfg['image_size']
+        heatmap_size = cfg['heatmap_size']
+        heatmaps = np.zeros((num_joints, heatmap_size[1], heatmap_size[0]),
+                            dtype='float32')
+        target_weight = np.zeros((num_joints, 1), dtype=np.float32)
+
+        for i in range(num_joints):
+            heatmap_vis = joints_3d_visible[i, 0]
```
Comment (innerlee) [inline mmpose/datasets/pipelines/top_down_transform.py:343]: `heatmap_vis` is not used very often in this function

File: mmpose/datasets/pipelines/top_down_transform.py [hunk 5]
```diff
@@ -307,14 +316,112 @@ def _generate_target(self, cfg, joints_3d, joints_3d_visible):
...
+                        target_weight,
+                        np.expand_dims(target_weight_i, axis=0)
+                    ],
+                                                   axis=0)
+            else:
+                target, target_weight = self._megvii_generate_target(
+                    results['ann_info'], joints_3d, joints_3d_visible,
+                    self.kernel)
+        else:
+            raise ValueError('Encoding approach %s is not supported!' %
```
Comment (innerlee) [inline mmpose/datasets/pipelines/top_down_transform.py:423]: f string

File: mmpose/models/keypoint_heads/top_down_mspn_head.py [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+        out_1 = out
+
+        out_2 = self.global_pooling(out_1)
+        out_2 = out_2.view(out_2.size(0), -1)
+        out_2 = self.middle_path(out_2)
+        out_2 = out_2.unsqueeze(2)
+        out_2 = out_2.unsqueeze(3)
+
+        out_3 = self.bottom_path(out_1)
+        out = out_1.mul(1 + out_2.expand_as(out_1).mul(out_3.expand_as(out_1)))
```
Comment (innerlee) [inline mmpose/models/keypoint_heads/top_down_mspn_head.py:124]: any reason to use `.mul` instead of `*`?

Comment (innerlee) [reply mmpose/models/keypoint_heads/top_down_mspn_head.py:124]: curios if the `expand_as` is needed, since `expand_as` need the tensors' shape support broadcasting.
will `out = out_1 * (1 + out_2) * out3` work? 
image

File: mmpose/models/keypoint_heads/top_down_mspn_head.py [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+
+        Returns:
+            out (list[Tensor]): a list of heatmaps from multiple stages.
+        """
+        out = []
+        assert isinstance(x, list)
+        assert len(x) == self.num_stages
+        for i in range(self.num_stages):
+            for j in range(self.num_units):
+                y = self.predict_layers[i * self.num_stages + j](x[i][j])
```
Comment (innerlee) [inline mmpose/models/keypoint_heads/top_down_mspn_head.py:181]: say if num_stages=2, num_units=3, then what will `i * self.num_stages + j` be?

image

File: mmpose/datasets/pipelines/top_down_transform.py [hunk 8]
```diff
@@ -307,14 +316,112 @@ def _generate_target(self, cfg, joints_3d, joints_3d_visible):
...
+                    dtype=np.float32)
+                target_weight = np.empty((0, num_joints, 1), dtype=np.float32)
+                for i in range(num_kernels):
+                    target_i, target_weight_i = self._megvii_generate_target(
+                        cfg, joints_3d, joints_3d_visible, self.kernel[i])
+                    target = np.concatenate(
+                        [target, np.expand_dims(target_i, axis=0)], axis=0)
+                    target_weight = np.concatenate([
+                        target_weight,
+                        np.expand_dims(target_weight_i, axis=0)
```
Comment (innerlee) [inline mmpose/datasets/pipelines/top_down_transform.py:415]: `np.expand_dims(target_weight_i, axis=0)` is `target_weight_i[None]`

File: mmpose/datasets/pipelines/top_down_transform.py [hunk 9]
```diff
@@ -307,14 +316,112 @@ def _generate_target(self, cfg, joints_3d, joints_3d_visible):
...
+        Returns:
+            tuple: A tuple containing targets.
+
+            - target: Target heatmaps.
+            - target_weight: (1: visible, 0: invisible)
+        """
+
+        num_joints = cfg['num_joints']
+        image_size = cfg['image_size']
+        heatmap_size = cfg['heatmap_size']
```
Comment (innerlee) [inline mmpose/datasets/pipelines/top_down_transform.py:337]: is heatmap size in `W, H` or `H, W`?

Comment (innerlee) [reply mmpose/datasets/pipelines/top_down_transform.py:337]: how about use `W, H = cfg['heatmap_size']` and use W and H in the following

File: mmpose/datasets/pipelines/top_down_transform.py [hunk 10]
```diff
@@ -307,14 +316,112 @@ def _generate_target(self, cfg, joints_3d, joints_3d_visible):
...
+
+            if (target_x >= heatmap_size[0] or target_x < 0) \
+                    or (target_y >= heatmap_size[1] or target_y < 0):
+                target_weight[i] = 0
+                continue
+
+            heatmaps[i, int(target_y), int(target_x)] = 1
+            heatmaps[i] = cv2.GaussianBlur(heatmaps[i], kernel, 0)
+            maxi = np.amax(heatmaps[i])
+            if maxi <= 1e-8:
```
Comment (innerlee) [inline mmpose/datasets/pipelines/top_down_transform.py:358]: in which case can `maxi <= 1e-8`?
from your code, before the blur, heatmap is all zero except for one point. isn't the maximum of heatmap after blur determined by kernel size? do you even need to actually take a maximum over it?

File: mmpose/models/backbones/rsn.py [hunk 11]
```diff
@@ -0,0 +1,603 @@
...
+    Local Representations for Multi-Person Pose Estimation" (ECCV 2020).
+
+    Args:
+        in_channels (int): Input channels of this block.
+        out_channels (int): Output channels of this block.
+        num_steps (int): Numbers of steps in RSB
+        stride (int): stride of the block. Default: 1
+        downsample (nn.Module): downsample operation on identity branch.
+            Default: None.
+        with_cp (bool): Use checkpoint or not. Using checkpoint will save some
```
Comment (innerlee) [inline mmpose/models/backbones/rsn.py:23]: the name `with_cp` feels so strange

File: mmpose/models/keypoint_heads/top_down_mspn_head.py (hunk 7 above)
Comment (wusize) [reply mmpose/models/keypoint_heads/top_down_mspn_head.py:181]: Sorry, this is a mistake. It's not found during training and testing because I set num_stages = 1. It should be i * self.num_units + j.

File: mmpose/datasets/pipelines/top_down_transform.py (hunk 8 above)
Comment (wusize) [reply mmpose/datasets/pipelines/top_down_transform.py:415]: OK. I will change it to target_weight_i[None].

File: mmpose/datasets/pipelines/top_down_transform.py (hunk 9 above)
Comment (wusize) [reply mmpose/datasets/pipelines/top_down_transform.py:337]: It's W,H in the config file. And ok, I'll change heatmap_size to W, H.

File: mmpose/datasets/pipelines/top_down_transform.py (hunk 10 above)
Comment (wusize) [reply mmpose/datasets/pipelines/top_down_transform.py:358]: This part was copied from the official RSN codes. I think they take the maximum because they needs to map the value range of the heatmap back to [0, 1] by dividing the blurred heatmap by maxi. In case of the divisor maxi being 0, they use if maxi <= 1e-8 then continue. 
I'll remove this part since maxi won't be 0 any way.

File: mmpose/models/backbones/rsn.py (hunk 11 above)
Comment (wusize) [reply mmpose/models/backbones/rsn.py:23]: I copied this from the official code of RSN. How about remove this arg? It's set False by default and seems never used.

File: mmpose/models/keypoint_heads/top_down_mspn_head.py (hunk 6 above)
Comment (wusize) [reply mmpose/models/keypoint_heads/top_down_mspn_head.py:124]: Yes, it works. I'll do it this way.

File: mmpose/datasets/pipelines/top_down_transform.py (hunk 4 above)
Comment (wusize) [reply mmpose/datasets/pipelines/top_down_transform.py:343]: OK. I'll use target_weight[i] = joints_3d_visible[i, 0] instead.

File: mmpose/datasets/pipelines/top_down_transform.py [hunk 12]
```diff
@@ -232,54 +242,49 @@ def _generate_target(self, cfg, joints_3d, joints_3d_visible):
         """
         num_joints = cfg['num_joints']
         image_size = cfg['image_size']
-        heatmap_size = cfg['heatmap_size']
+        [W, H] = cfg['heatmap_size']
```
Comment (innerlee) [inline mmpose/datasets/pipelines/top_down_transform.py:245]: `W, H = cfg['heatmap_size']`

File: mmpose/datasets/pipelines/top_down_transform.py [hunk 13]
```diff
@@ -307,14 +311,101 @@ def _generate_target(self, cfg, joints_3d, joints_3d_visible):
...
+            target_x = joints_3d[i, 0] * W / image_size[0]
+
+            if (target_x >= W or target_x < 0) \
+                    or (target_y >= H or target_y < 0):
+                target_weight[i] = 0
+                continue
+
+            heatmaps[i, int(target_y), int(target_x)] = 1
+            heatmaps[i] = cv2.GaussianBlur(heatmaps[i], kernel, 0)
+            maxi = np.amax(heatmaps[i])
```
Comment (innerlee) [inline mmpose/datasets/pipelines/top_down_transform.py:351]: does `maxi = heatmaps[i, int(target_y), int(target_x)]`?

File: mmpose/models/backbones/rsn.py [hunk 14]
```diff
@@ -0,0 +1,591 @@
...
+            outputs_i = list()
+            outputs.append(outputs_i)
+            for j in range(i + 1):
+                if j == 0:
+                    inputs = spx[i]
+                else:
+                    inputs = outputs[i][j - 1]
+                if i > j:
+                    inputs += outputs[i - 1][j]
+                module_name = 'conv_bn_relu2_' + str(i + 1) + '_' + str(j + 1)
```
Comment (innerlee) [inline mmpose/models/backbones/rsn.py:103]: use f-string

File: mmpose/models/backbones/rsn.py [hunk 15]
```diff
@@ -0,0 +1,591 @@
...
+        self.num_steps = num_steps
+        self.norm_cfg = norm_cfg
+        self.layer1 = self._make_layer(
+            block,
+            in_channels,
+            num_blocks[0],
+            expand_times=expand_times,
+            res_top_channels=in_channels)
+        for i in range(1, num_units):
+            module_name = 'layer' + str(i + 1)
```
Comment (innerlee) [inline mmpose/models/backbones/rsn.py:162]: f-string

File: mmpose/models/keypoint_heads/top_down_mspn_head.py [hunk 16]
```diff
@@ -0,0 +1,196 @@
...
+    """
+
+    def __init__(self,
+                 out_shape,
+                 unit_channels=256,
+                 out_channels=17,
+                 num_stages=4,
+                 num_units=4,
+                 use_prm=False,
+                 norm_cfg=dict(type='BN')):
```
Comment (innerlee) [inline mmpose/models/keypoint_heads/top_down_mspn_head.py:152]: when using dict as default arg, pls protect it in the beginning of the function

```python
# Protect mutable default arguments
norm_cfg = cp.deepcopy(norm_cfg)
```

normally you think your code wouldn't wrong. however once it bugs, it will be extremely hard to debug

File: mmpose/datasets/pipelines/top_down_transform.py (hunk 13 above)
Comment (wusize) [reply mmpose/datasets/pipelines/top_down_transform.py:351]: Yes. But heatmaps[i, int(target_y), int(target_x)] will be less than 1 after being gaussian blurred. So we divide heatmaps by maxi to keep the max value = 1.
2020-11-06 10-43-53屏幕截图

Comment (innerlee) [reply mmpose/datasets/pipelines/top_down_transform.py:351]: we can save one pass

**Comment by innerlee** [comment]
tests are failing

File: mmpose/datasets/pipelines/top_down_transform.py (hunk 13 above)
Comment (innerlee) [reply mmpose/datasets/pipelines/top_down_transform.py:351]: other than the amax, if you can pre-scale the gaussian kernel, then more cpu time could be saved

image

Comment (innerlee) [reply mmpose/datasets/pipelines/top_down_transform.py:351]: note this is not required. optimization can always come back later

Comment (wusize) [reply mmpose/datasets/pipelines/top_down_transform.py:351]: We can use maxi = heatmaps[i, int(target_y), int(target_x)]  instead of np.amax.

**Comment by innerlee** [comment]
is this still block by accuracy?

File: mmpose/datasets/pipelines/top_down_transform.py [hunk 17]
```diff
@@ -232,54 +242,49 @@ def _generate_target(self, cfg, joints_3d, joints_3d_visible):
...
         joint_weights = cfg['joint_weights']
         use_different_joint_weights = cfg['use_different_joint_weights']
 
         target_weight = np.zeros((num_joints, 1), dtype=np.float32)
-        target = np.zeros((num_joints, heatmap_size[1], heatmap_size[0]),
-                          dtype=np.float32)
+        target = np.zeros((num_joints, H, W), dtype=np.float32)
 
-        tmp_size = self.sigma * 3
+        tmp_size = sigma * 3
```
Comment (innerlee) [inline mmpose/datasets/pipelines/top_down_transform.py:252]: `tmp_size` is a vague name, `3` is a magic number. Could you add a line of comment?

Comment (wusize) [reply mmpose/datasets/pipelines/top_down_transform.py:252]: Ok.

Comment (wusize) [reply mmpose/datasets/pipelines/top_down_transform.py:252]: For a gaussian distribution, more than 99% of the energy is contained within the range [mu-3*sigma, mu +3*sigma].
2020-11-09 18-18-28屏幕截图

**Comment by innerlee** [comment]
tests are failing

File: mmpose/datasets/pipelines/top_down_transform.py [hunk 18]
```diff
@@ -307,14 +311,99 @@ def _generate_target(self, cfg, joints_3d, joints_3d_visible):
...
+        W, H = cfg['heatmap_size']
+        heatmaps = np.zeros((num_joints, H, W), dtype='float32')
+        target_weight = np.zeros((num_joints, 1), dtype=np.float32)
+
+        for i in range(num_joints):
+            target_weight[i] = joints_3d_visible[i, 0]
+
+            if target_weight[i] < 1:
+                continue
+            target_y = joints_3d[i, 1] * H / image_size[1]
```
Comment (innerlee) [inline mmpose/datasets/pipelines/top_down_transform.py:341]: make target_x  and target_y int here

**Review by innerlee — APPROVED** [review]

**Comment by innerlee** [comment]
looks good after fixing the ci
